### PR TITLE
Sublime v1 collateral ratio

### PR DESCRIPTION
**Pull request by aditya26026**
Added an extra function so that the stack does not get too deep.

### CODE REVIEW

**Comment by prateekreddy** [comment]
Copy the storage variable into memory, if you are going to use it multiple times.

**Review by prateekreddy — CHANGES_REQUESTED** [review]

File: contracts/Pool/Pool.sol [hunk 1]
```diff
@@ -354,18 +354,101 @@ contract Pool is ERC20PresetMinterPauserUpgradeable,IPool {
...
+                collateralAsset
+            );
+
+        uint256 _ratioOfPrices =
+            IPriceOracle(IPoolFactory(PoolFactory).priceOracle())
+                .getLatestPrice(collateralAsset, borrowAsset);
+
+        (uint256 _interest, uint256 _gracePeriodsTaken) =
+            (
+                IRepayment(Repayment).calculateRepayAmount(
```
[the remark this answers is not in the document]
Comment (aditya26026) [reply contracts/Pool/Pool.sol:370]: This has to be done

File: contracts/Pool/Pool.sol [hunk 2]
```diff
@@ -354,18 +354,101 @@ contract Pool is ERC20PresetMinterPauserUpgradeable,IPool {
...
+            gracePenaltyRate.mul(amountPerPeriod()).div(100).mul(
+                _gracePeriodsTaken
+            )
+        );
+        if (_interest < _extraInterest) {
+            _interest = 0;
+        } else {
+            _interest = _interest.sub(_extraInterest);
+        }
+        _ratio = (_currentCollateralTokens.mul(_ratioOfPrices)).div(
```
[the remark this answers is not in the document]
Comment (aditya26026) [reply contracts/Pool/Pool.sol]: ChainLink multiplies the ratios by a factor of 10^8. So I have divided the numbers by the same

File: contracts/Pool/Pool.sol [hunk 3]
```diff
@@ -62,7 +62,7 @@ contract Pool is ERC20PresetMinterPauserUpgradeable,IPool {
     uint256 public extensionVoteEndTime;
     uint256 public noOfGracePeriodsTaken;
     uint256 nextDuePeriod;
-
+    uint256 gracePenaltyRate;
```
[the remark this answers is not in the document]
Comment (aditya26026) [reply contracts/Pool/Pool.sol]: done

File: contracts/Pool/Pool.sol [hunk 4]
```diff
@@ -354,18 +354,101 @@ contract Pool is ERC20PresetMinterPauserUpgradeable,IPool {
     event Liquidated(address liquidator, address lender);
 
     // TODO
```
[the remark this answers is not in the document]
Comment (aditya26026) [reply contracts/Pool/Pool.sol]: done

File: contracts/Pool/Pool.sol [hunk 5]
```diff
@@ -61,8 +61,8 @@ contract Pool is ERC20PresetMinterPauserUpgradeable,IPool {
     mapping(address => LendingDetails) public lenders;
     uint256 public extensionVoteEndTime;
     uint256 public noOfGracePeriodsTaken;
-    uint256 nextDuePeriod;
-
+    uint256 public nextDuePeriod;
+    uint256 public gracePeriodPenaltyFraction;
```
Comment (prateekreddy) [inline contracts/Pool/Pool.sol:65]: This will be part of poolFactory

File: contracts/Pool/Pool.sol [hunk 6]
```diff
@@ -353,19 +353,82 @@ contract Pool is ERC20PresetMinterPauserUpgradeable,IPool {
...
+                ((calculateCurrentPeriod().add(1)).mul(_repaymentLength))
+                    .add(_loanStartedAt)
+                    .sub(block.timestamp)
+            );
+        _interest = _interest.sub(
+            gracePeriodPenaltyFraction.mul(_interestPerPeriod).div(100).mul(
+                _gracePeriodsTaken
+            )
+        );
+        if (_interest < _extraInterest) {
```
Comment (prateekreddy) [inline contracts/Pool/Pool.sol:384]: Is this possible ? Does interest already include the extraInterest as well right

File: contracts/Pool/Pool.sol [hunk 7]
```diff
@@ -353,19 +353,82 @@ contract Pool is ERC20PresetMinterPauserUpgradeable,IPool {
...
-    // TODO
     function getCurrentCollateralRatio(address _lender)
         public
-        returns (uint256 ratio) {
-
+        returns (uint256 _ratio)
+    {
+        uint256 _balanceOfLender = balanceOf(_lender);
+        uint256 _totalSupply = totalSupply();
+        address _collateralToken = collateralAsset;
```
Comment (prateekreddy) [inline contracts/Pool/Pool.sol:418]: If both use similar code, then can you seperate the common code into a different internal function

Comment (prateekreddy) [reply contracts/Pool/Pool.sol:418]: I mean getCurrentCollateralRatio

**Review by prateekreddy — CHANGES_REQUESTED** [review]

File: contracts/Pool/Pool.sol (hunk 7 above)
Comment (aditya26026) [reply contracts/Pool/Pool.sol:418]: will do

File: contracts/Pool/Pool.sol (hunk 6 above)
Comment (aditya26026) [reply contracts/Pool/Pool.sol:384]: Say he has to pay 100 dai in 8 days. The borrower has already paid 60 dai for an interval. The _interest left to be paid would be 40 dai for that interval. On the 4th day, the _extraInterest would be 50 dai. So in that case the borrower does not owe money to the pool.

File: contracts/Pool/Pool.sol (hunk 5 above)
Comment (aditya26026) [reply contracts/Pool/Pool.sol:65]: The problem with having this as a global variable is that changing this during the execution of a pool would mess up all the calculations for liquidation and repayments.

Comment (prateekreddy) [reply contracts/Pool/Pool.sol:65]: ok sure, will intialize the pool with the value, in that case

File: contracts/Pool/Pool.sol (hunk 6 above)
Comment (prateekreddy) [reply contracts/Pool/Pool.sol:398]: cool, makes sense.

**Review by prateekreddy — APPROVED** [review]

**Pull request merged**